File: matlab/River.m
classdef River
    
    properties
        length
        basePointsL
        basePointsU
        minWaterUnderKeel
        
        streamVel
        waterDepth
        PointDataset
        PointCoordinates
    end
    
    methods
        function o = River()
            o.minWaterUnderKeel = 0.2;
            
            %load river data from inp file
            
            % river part from km 655 - 665
%             opts=detectImportOptions('MW_grid_500m_655_665_UTM32.inp','FileType','text');
%             data2 = readmatrix('MW_grid_500m_655_665_UTM32.inp',opts);

        % whole river
            opts=detectImportOptions('MW_grid_500m_655_852_UTM32.inp','FileType','text');
            data2 = readmatrix('MW_grid_500m_655_852_UTM32.inp',opts);
            
            index_ones = find(data2(:,1)==1);
            o.PointCoordinates = data2(index_ones(1):index_ones(2)-9,2:3); %Matrix with all point coordinates
            o.PointDataset = data2(index_ones(2):end,2:7); % Matrix with all point dataset (like stream velocity,...)
            
            o.streamVel = zeros(length(o.PointCoordinates),1);
            o.waterDepth = zeros(length(o.PointCoordinates),1);

            for i=1:length(o.PointCoordinates)
                o.streamVel(i) = sqrt(o.PointDataset(i,1)^2 + o.PointDataset(i,2)^2);
                o.waterDepth(i) = o.PointDataset(i,3);
            end
            
            o.streamVel  = reshape(o.streamVel,26,[]);
            o.waterDepth  = reshape(o.waterDepth,26,[]);            
            o.PointDataset = reshape(o.PointDataset,26,[],6);
            o.PointCoordinates = reshape(o.PointCoordinates,26,[],2);
            
            
            
            o.length = 20*length(o.PointCoordinates);
            
            %%%%%%%% CREATE BASEPOINTs %%%%%%%%
            
            % X VALUE rand distr with specific mean dist
            meanBasePointDist = 500;
            %             o.basePointsL = [0 sort(randperm(o.length,round(o.length/meanBasePointDist))) o.length];
            %             o.basePointsU = [0 sort(randperm(o.length,round(o.length/meanBasePointDist))) o.length];
            
            o.basePointsL = (0:meanBasePointDist:o.length);
            o.basePointsL = o.basePointsL + normrnd(0,10,[1 length(o.basePointsL)]);
            o.basePointsU = (0:meanBasePointDist:o.length);
            o.basePointsU = o.basePointsU + normrnd(0,10,[1 length(o.basePointsU)]);
            o.basePointsL(2,:) = 0;
            o.basePointsU(2,:) = 0;
            
            
            % Y VALUE with AR1 process
            AR1 = 20;
            for i=1:length(o.basePointsU)
                AR1 = 0.0100 + AR1 *  0.9995 + normrnd(0,sqrt( 0.3998));
                
                %                 if isber(i,o.basePointsL(1,:))
                %                     o.membasePointsL(2,find(o.basePointsL(1,:) == i)) = AR1 - 20 - randi([20 40]);
                temp = 55;
                o.basePointsL(2,i) = 250-randi(temp*[1 1]);
                %                 elseif ismember(i,o.basePointsU(1,:))
                %                     o.basePointsU(2,find(o.basePointsU(1,:) == i)) = AR1 - 20 + randi([20 40]);
                o.basePointsU(2,i) = 250+randi(temp*[1 1]);
                %                 end
            end
            
        end
        
        function   [xy_utm] = getPosUTMfromPos(o,xy)
            x = xy(1);
            y = xy(2);
            x_log = floor(x/20);
            xAlpha = x/20 - x_log;
            y_log = floor(y/20)+1;
            yAlpha = y/20 + 1 - y_log; 
            
            base_utm = o.PointCoordinates(y_log,x_log,:);
            
            % in x direction           
            diff = o.PointCoordinates(y_log,x_log + 1,:) - o.PointCoordinates(y_log,x_log,:);
            
            base_utm(:,:,1) = base_utm(:,:,1) + xAlpha * diff(:,:,1) ;
            base_utm(:,:,2) = base_utm(:,:,2) + xAlpha * diff(:,:,2) ;
            
            % in y direction           
            diff = o.PointCoordinates(y_log + 1,x_log,:) - o.PointCoordinates(y_log,x_log,:);
            x_utm = base_utm(:,:,1) + yAlpha * diff(:,:,1) ;
            y_utm = base_utm(:,:,2) + yAlpha * diff(:,:,2) ;
           
           xy_utm = [y_utm;x_utm];
            
                       
            
        end
        
        function  basePointsL = getClosestBasePointsL(o,x,dir,num)
            
            if dir == 1
                index = find(o.basePointsL(1,:) > x,1)-1;
            else
                index = find(o.basePointsL(1,:) > x,1);
            end
            isDone = 0;
            basePointsL = [];
            i = index;
            while ~isDone
                basePointsL = [basePointsL o.basePointsL(:,i)];
                i = i + dir;
                if abs(i - (index)) == num
                    isDone = 1;
                end
            end
        end
        
        function  basePointsU = getClosestBasePointsU(o,x,dir,num)
            
            if dir == 1
                index = find(o.basePointsU(1,:) > x,1)-1;
            else
                index = find(o.basePointsU(1,:) > x,1);
            end
            isDone = 0;
            basePointsU = [];
            i = index;
            while ~isDone
                basePointsU = [basePointsU o.basePointsU(:,i)];
                i = i + dir;
                if abs(i - (index)) == num
                    isDone = 1;
                end
            end
            
        end
        
        function  vStream = getMeanVStreamForVessel(o,ships, ID)
            
            minY_log = round((ships.y(ID)-ships.width_eff(ID)/2)/20);
            maxY_log = round((ships.y(ID)+ships.width_eff(ID)/2)/20);
            minX_log = round((ships.x(ID)-ships.length(ID)/2)/20)+10;
            maxX_log = round((ships.x(ID)+ships.length(ID)/2)/20)+10;
                        
            vStream = ships.dir(ID) *  mean(o.streamVel(minY_log:maxY_log,minX_log:maxX_log),'all');
%              vStream = 2;
        end
        
        function A = getRiverProfileForVessel(o,ships, ID)
            minX_log = round((ships.x(ID)-ships.length(ID)/2)/20)-10;
            maxX_log = round((ships.x(ID)+ships.length(ID)/2)/20)+10;
            
            n = maxX_log - minX_log;
            A = mean(trapz(o.waterDepth(:,minX_log:maxX_log))) * 20;

%             A = 1500;
        end

        function h = getWaterDepthForVessel(o,ships,ID)
            minX_log = round((ships.x(ID)-ships.length(ID)/2)/20)-10;
            maxX_log = round((ships.x(ID)+ships.length(ID)/2)/20)+10;
            minY_log = floor(ships.y(ID)/20);
            maxY_log = ceil(ships.y(ID)/20);
            
            h = mean(mean(o.waterDepth(minY_log:maxY_log,minX_log:maxX_log)));
%             h = 5;
        end
        function [lowerPoints, upperPoints] = detectFreeSpace(o,ID,ships,lookAheadDistance,numPoints)
            
            lookAheadDistance = lookAheadDistance/20;
            posX = round((ships.x(ID)-ships.length(ID) / 2)/20);
            ssq = ships.ssq(ID);
            
            lengthBlock = lookAheadDistance/numPoints;
            if ships.dir(ID) == 1
                start = round((floor(posX/lengthBlock)-1)*lengthBlock);
            else
                start = round((ceil(posX/lengthBlock)+1)*lengthBlock) - lookAheadDistance+1;
            end
            waterDepth = o.waterDepth(:,start:start + lookAheadDistance-1);              
            
            
            
            freeSpace = waterDepth -  (ssq + o.minWaterUnderKeel) < 0; % binary array ( 1 -> coast ; 0 -> freeSpace)
            
            % consider vessel length 
            for i=1:round(ships.length(ID)/20)
            freeSpace = [freeSpace zeros(26,1)] | [zeros(26,1) freeSpace];
            end
            
            lowerPoly = [];
            upperPoly = [];
            for i=1:size(freeSpace,2)
                [~,b]= max(diff(find(diff([1; freeSpace(:,i); 1])~=0)));
                try
                index = b(1);
                catch
                     disp a
                end
                positions = find(diff([1; freeSpace(:,i); 1])~=0)-1;
                lowerPoly(i) = positions(index);
                upperPoly(i) =  positions(index +1);
                
            end
            
            % divide freeSpace into numPoints blocks and find maximum
            
            blocks =  round(linspace(1,length(upperPoly),numPoints+1));
            blocks2 =  round(linspace(1,length(lowerPoly),numPoints+1));
            for i=1:numPoints
                [x,y] = min(upperPoly(blocks(i):blocks(i+1)));
                upperPoints(:,i) = [blocks(i)+y-1;x];
                [x,y] = max(lowerPoly(blocks2(i):blocks2(i+1)));
                lowerPoints(:,i) = [blocks2(i)+y-1;x];
            end
            
%             for i=1:numPoints
%                 [x,y] = min(upperPoly(blocks(i):blocks(i)));
%                 upperPoints(:,i) = [blocks(i)+y-1;x];
%                 [x,y] = max(lowerPoly(blocks2(i):blocks2(i)));
%                 lowerPoints(:,i) = [blocks2(i)+y-1;x];
%             end
            
                upperPoints = [(start-1+ upperPoints(1,:))*20;upperPoints(2,:) *20]';
                lowerPoints = [(start-1+ lowerPoints(1,:))*20;lowerPoints(2,:) *20]';

            
            
            
%             [~,upperPoly_binary] = max(freeSpace==0,[],1);
%             [~,lowerPoly_binary] = max(flip(freeSpace)==0,[],1);
            
            
            % fill gaps in polyline until reach numPoints
%             y1 = upperPoly_binary(1);
%             upperPoly(:,1) = [1;y1];
%             for i=2:length(upperPoly_binary)
%                 y2 = upperPoly_binary(i);
%                 if y2 ~= y1
%                     upperPoly = [upperPoly [i-1;y1]];
%                     upperPoly = [upperPoly [i;y2]];
%                 end
%                 y1 = y2;                
%             end
%             upperPoly = [upperPoly [length(upperPoly_binary);upperPoly_binary(end)]];
%             upperPoly
%             
%             isDone = 0;
%             gap = 1;
%             while ~isDone
%                 i = 0;
%                 while i <= length(upperPoly)-2
%                     
%                     if i == 0
%                         y1 = Inf;
%                     else
%                         y1 = upperPoly(2,i);
%                     end
%                     if i == length(upperPoly)-2
%                         y2 = Inf;
%                     else
%                         y2 = upperPoly(2,i+3);
%                     end
%                     
%                     if upperPoly(1,i+2)-upperPoly(1,i+1) <= gap && upperPoly(2,i+1) < y1  && upperPoly(2,i+1) < y2
%                         if i ~= 0
%                         upperPoly(1,i) = upperPoly(1,i)+1;
%                         end
%                         if i+3 <= length(upperPoly)
%                             upperPoly(1,i+3) = upperPoly(1,i+3)-1;
%                         end
%                         upperPoly(:,i+1:i+2) = [];
%                         if length(upperPoly) <= numPoints
%                             isDone = 1;
%                             break;
%                         end
%                         if i~=0 && i+1 <= length(upperPoly)
%                             if upperPoly(2,i) == upperPoly(2,i+1)
%                                 upperPoly(:,i:i+1) = [];
%                             end
%                         end
%                         if length(upperPoly) <= numPoints
%                             isDone = 1;
%                             break;
%                         end
%                         
%                         
%                     else
%                     i = i+2;
%                     end
% 
%                 end
%                 gap = gap + 1;
%             end
            
            
            
%             lowerPoly = 26-lowerPoly+1;
            
            
    
            
        end
    end
    
end

File: python/river.py
"""
This will be a Docstring in the future

"""

from numpy.core.numeric import indices
import pandas as pd
import numpy as np 


class River:
    def __init__(self) -> None:
        
        # Number of evenly spaced gridpoints for which stream velocity and water depth data is available. 
        # All Gridpoints are BASEPOINT_DIST meters apart to span a maximum rhine width of 500 meters 
        self.GRIDPOINTS_PER_WIDTH = 26
        self.BASEPOINT_DIST = 20

        self.min_water_under_keel = 0.2

        # Coordninates for the river loctation
        self.point_coords = pd.read_csv("python/data/MW_grid_500m_655_852_UTM32_1.txt", sep=" ", skiprows=1)
        self.point_coords = self.point_coords.iloc[:,1:3] # Only longitudinal and lateral datapoints

        # Data for water depth and stream velocity
        self.point_dataset = pd.read_csv("python/data/MW_grid_500m_655_852_UTM32_2.txt", sep=" ")
        self.point_dataset = self.point_dataset.iloc[:,0:6]

        # Init vectors for water depth and stream velocity
        self.water_depth = self.stream_vel = np.zeros((len(self.point_coords)))

        self.stream_vel = np.sqrt(np.add(self.point_dataset.iloc[:,0].values**2, self.point_dataset.iloc[:,1].values**2))
        self.water_depth = self.point_dataset.iloc[:,2].values

        # Reshape stream velocity and water depth to have GRIDPOINTS_PER_WIDTH columns
        self.stream_vel = self.stream_vel.reshape((-1, self.GRIDPOINTS_PER_WIDTH))
        self.water_depth = self.water_depth.reshape((-1, self.GRIDPOINTS_PER_WIDTH))

        self.point_coords = self.point_coords.values.reshape((-1,self.GRIDPOINTS_PER_WIDTH,2))        
        self.point_dataset = self.point_dataset.values.reshape((-1,self.GRIDPOINTS_PER_WIDTH,6))

        # Create upper and lower basepoints

        self.length = self.BASEPOINT_DIST * len(self.point_coords)

        mean_base_point_dist =  500

        # Create equally spaces points for the entire river length, and add some random noise to them.
        self.base_points_l = np.arange(start=0,stop=self.length,step=mean_base_point_dist)
        self.base_points_l = np.add(self.base_points_l,np.random.normal(loc=0,scale=10,size=len(self.base_points_l)))

        self.base_points_u = np.arange(start=0,stop=self.length,step=mean_base_point_dist)
        self.base_points_u = np.add(self.base_points_u, np.random.normal(loc=0,scale=10,size=len(self.base_points_u)))

        # Add second dimension with all zeros
        zeros = np.zeros(len(self.base_points_u))
        self.base_points_l = np.append(self.base_points_l, zeros).reshape(2,-1)
        self.base_points_u = np.append(self.base_points_u, zeros).reshape(2,-1)

    def get_utm_position(self, x, y):

        x_log = int(x/self.BASEPOINT_DIST) # Casting to int, i.e. flooring
        x_alpha = x/self.BASEPOINT_DIST - x_log
        
        y_log = int(y/self.BASEPOINT_DIST) + 1 # Casting to int, i.e. flooring
        y_alpha = y/self.BASEPOINT_DIST + 1 - y_log

        # Base coordinates after utm transform
        base_utm = self.point_coords[x_log,y_log,:]

        # Difference between points in x direction
        diff_x = self.point_coords[x_log+1,y_log,:] - base_utm

        base_utm += x_alpha * diff_x

        # Difference between points in y direction
        diff_y = self.point_coords[x_log,y_log+1,:] - base_utm

        x_utm = base_utm[0] + y_alpha * diff_y[0]
        y_utm = base_utm[1] + y_alpha * diff_y[1]

        return [x_utm, y_utm]

        
    # This is the same function as in River.m:148-157 but vectorized. 
    # Therefore mean stream velocities for every vessel are calculated at once.
    # This means the function only gets called once per timestep and not once per ship per timestep
    def mean_stream_vel(self, ships):

        min_y_log = np.round((ships.y_location - ships.eff_width / 2) / self.BASEPOINT_DIST).astype(int)
        max_y_log = np.round((ships.y_location + ships.eff_width / 2) / self.BASEPOINT_DIST).astype(int)

        min_x_log = np.round((ships.x_location - ships.eff_width / 2) / self.BASEPOINT_DIST).astype(int) + 10
        max_x_log = np.round((ships.x_location + ships.eff_width / 2) / self.BASEPOINT_DIST).astype(int) + 10

        # Define empty array for all mean stream velocities to be written into
        v_stream = np.zeros(len(min_y_log))
        for i in range(len(v_stream)):
            v_stream[i] = ships.direction[i] *\
                 np.mean(self.stream_vel[min_x_log[i]:max_x_log[i]+1, min_y_log[i]:max_y_log[i]+1])

        return v_stream

    # Vectorized function for receiving the current water depth per vessel (orig: River.m:169)
    def get_water_depth(self,ships):

        min_x_log = np.round((ships.x_location - ships.length / 2) / self.BASEPOINT_DIST).astype(int) - 10
        max_x_log = np.round((ships.x_location + ships.length / 2) / self.BASEPOINT_DIST).astype(int) + 10

        min_y_log = np.floor(ships.y_location / self.BASEPOINT_DIST).astype(int)
        max_y_log = np.ceil(ships.y_location / self.BASEPOINT_DIST).astype(int)

        depth = np.zeros(ships.num_ships)

        for i in range(len(depth)):
            depth[i] = np.mean(np.mean(self.water_depth[min_x_log[i]:max_x_log[i]+1,min_y_log[i]:max_y_log[i]+1]))

        return depth

    # Find the river profile for every ship (orig River.m:159)
    def get_river_profile(self,ships):

        min_x_log = np.round((ships.x_location - ships.length / 2) / self.BASEPOINT_DIST).astype(int) - 10
        max_x_log = np.round((ships.x_location + ships.length / 2) / self.BASEPOINT_DIST).astype(int) + 10

        river_profile = np.zeros(ships.num_ships)

        for i in range(len(river_profile)):
            river_profile[i] = np.mean(np.trapz(self.water_depth[min_x_log[i]:max_x_log[i]+1,:])) * self.BASEPOINT_DIST

        return river_profile

    # Vectorized version of the free space under keel function in River.m:178
    def detect_free_space(self, ID, ships, look_ahead_dist, num_points):

        look_ahead_dist = look_ahead_dist / self.BASEPOINT_DIST

        # Get x position of vessel by subtracting its length from its current x coordinate
        pos_x = np.round((ships.x_location[ID] - ships.length[ID] / 2) / self.BASEPOINT_DIST)

        # Get vessel squat values
        ssq = ships.squat[ID]

        # Split the lookahead into num_points blocks for lateral control policy
        length_block = look_ahead_dist / num_points


        if ships.direction[ID] == 1:
            start = np.round((np.floor(pos_x/length_block) - 1) * length_block)
        else:
            start = np.round((np.ceil(pos_x/length_block) + 1) * length_block) - look_ahead_dist + 1
        
        # This returns a list with num_ships arrays containing the respective water depths [water_depth1, water_depth2,...]
        water_depth_array  = self.water_depth[int(start)-1:(int(start) + int(look_ahead_dist) -1), :]

        # List containing logical arrays per ship with 1 being coast and 0 being free space
        free_space = water_depth_array - (ssq + self.min_water_under_keel) < 0

        # Extent the coast indicator points (ones) per length-block over the length of the vessel
        # to avoid it to reeve into the coast because there is clearance in front
        #
        # This is achieved by duplicating the freespace matrix and adding one column of zeros 
        # to the beginning of the first and to the end of the second. This moves the first matrix one step in x direction
        # Now both matrices are iteratively XOR-ed to carry over high points (ones) from one step to the next, 
        # thereby extending high points of the matix over time.
        for i in range(int(np.round(ships.length[ID] / self.BASEPOINT_DIST))):
            a = np.r_[free_space, [np.zeros(free_space.shape[1])]]
            b = np.r_[[np.zeros(free_space.shape[1])], free_space]
            free_space = np.array(a, dtype=bool) | np.array(b, dtype=bool)

        # Find the first points in the logic freespace array that belong to the coast
        # Freespace array: 26xN
        # |1 1 1 1 1 1 1 1 1 1| > y1
        # |0 1 0 0 1 0 0 0 0 1| > y ..
        # |0 0 0 0 0 0 0 1 0 0| > y ..
        # |1 1 1 1 1 1 1 1 1 1| > y_26
        # - dir of vessel (N) ->
        # 
        # Here the min and max y-indices that contain a transision from 0 to one are returned
        # to determine where the vessel can safely drive
        #
        lower_poly = np.zeros(free_space.shape[0])
        upper_poly = np.zeros(free_space.shape[0])

        for i in range(free_space.shape[0]):
            added_ones = np.insert(free_space[i], [0,len(free_space[i])], 1).astype(np.float32)
            diff_of_vals = np.abs(np.diff(added_ones).astype(np.float32))
            non_zero_indices = np.where((diff_of_vals != 0))
            diff_of_indices = np.diff(non_zero_indices)
            max_index = np.argmax(diff_of_indices)


            lower_poly[i] = non_zero_indices[0][max_index]
            upper_poly[i] = non_zero_indices[0][max_index + 1]

        # From here on the results of the matlab implementation differ from this one due to zero indexing.
        # The linspacer starts at one as it does in the original file River.m:218.
        # However, the first one-index is used to index the first element of the upper_poly vector in l. 203.
        # In matlab this works due to one-indexing. In python it does not. Therefore the first element in 
        # the upper_poly array is omitted. If the linspacer would start at zero, the breakpoints of the spacer
        # would differ from the original. Difference is minor, therefore no further adjustments are made.
        blocks_upper = np.round(np.linspace(1,len(upper_poly),num_points + 1)).astype(int)
        blocks_lower = np.round(np.linspace(1,len(upper_poly),num_points + 1)).astype(int)

        upper_points = np.zeros((num_points,2))
        lower_points = np.zeros((num_points,2))

        for i in range(num_points):
            
            current_upper_range = upper_poly[blocks_upper[i]:blocks_upper[i+1]]
            current_lower_range = lower_poly[blocks_lower[i]:blocks_lower[i+1]]

            upper_min = np.min(current_upper_range)
            upper_ind = np.argmin(current_upper_range) 
            upper_points[i,:] = upper_ind + blocks_upper[i], upper_min

            lower_max = np.max(current_lower_range)
            lower_ind = np.argmax(current_lower_range)
            lower_points[i,:] = lower_ind + blocks_lower[i], lower_max

        upper_points[:,0] = (upper_points[:,0] + start -1) * self.BASEPOINT_DIST
        upper_points[:,1] = (upper_points[:,1]) * self.BASEPOINT_DIST
        lower_points[:,0] = (lower_points[:,0] + start -1) * self.BASEPOINT_DIST
        lower_points[:,1] = (lower_points[:,1]) * self.BASEPOINT_DIST

        return (lower_points, upper_points)
